File: Thursday/mm.py
import random
import time
 
def print_mat(M):
        for r in M:
                print r
        print '\n\n'

# Set up some random matrices.
a = []
b = []
c = []
N = 3
for i in range(N):
                a.append([random.randint(1, 4) for i in range(N)])
                b.append([random.randint(1, 4) for i in range(N)])
 
# Run the code in the non-permuted way
num_iters = 1
 
A = []
B = []
C = []
for i in range(N):
                A.append(a[i])
                B.append(b[i])
                C.append([0 for i in range(N)])
 
start = time.time()
for it in range(num_iters):
                for row in range(N):
                                for col in range(N):
                                                for k in range(N):
                                                                C[row][col] += (A[row][k]*B[k][col])
end = time.time()
print end - start
print_mat(A)
print_mat(B)
print_mat(C)
A = []
B = []
C = []

for i in range(N):
                A.append(a[i])
                B.append(b[i])
                C.append([0 for i in range(N)])
# run the code in the permuted way
start = time.time()
for it in range(num_iters):
                for k in range(N):
                                for col in range(N):
                                                for row in range(N):
                                                                C[row][col] += (A[row][k]*B[k][col])
end = time.time()
print end - start

print_mat(A)
print_mat(B)
print_mat(C)

File: Thursday/mm.cpp
#include <iostream>
#include <ctime>
 
using namespace std;
 
int main() {
       const int N = 4;
       const int num_iters = 100000;
       double A[N][N], a[N][N];
       double B[N][N], b[N][N];
       double C[N][N];
 
       for (int r = 0; r < N; r++){
              for (int c = 0; c < N; c++){
                     a[r][c] = (double)(rand() % 5);
                     b[r][c] = (double)(rand() % 5);
              }
       }
       for (int r = 0; r < N; r++){
              for (int c = 0; c < N; c++){
                     A[r][c] = a[r][c];
                     B[r][c] = b[r][c];
                     C[r][c] = 0;
              }
       }
       /*
       for (int r = 0; r < N; r++){
              for (int c = 0; c < N; c++){
                     cout << A[r][c] << " ";
              }
              cout << endl;
       }
 
 
       for (int r = 0; r < N; r++){
              for (int c = 0; c < N; c++){
                     cout << B[r][c] << " ";
              }
              cout << endl;
       }
       */
       clock_t start, end;
       start = clock();
       for (int iter = 0; iter < num_iters; iter++){
              for (int c = 0; c < N; c++){
                     for (int r = 0; r < N; r++){
                           for (int k = 0; k < N; k++){
                                  C[r][c] += (A[r][k] * B[k][c]);
                           }
                     }
              }
       }
       end = clock();
      
       for (int r = 0; r < N; r++){
              for (int c = 0; c < N; c++){
                     cout << C[r][c] << " ";
              }
              cout << endl;
       }
      
       cout << "The time taken to run this code was " << (double)(end - start)/CLOCKS_PER_SEC << endl;
 
       for (int r = 0; r < N; r++){
              for (int c = 0; c < N; c++){
                     A[r][c] = a[r][c];
                     B[r][c] = b[r][c];
                     C[r][c] = 0;
              }
       }
       start = clock();
       for (int iter = 0; iter < num_iters; iter++){
              for (int c = 0; c < N; c++){
              for (int k = 0; k < N; k++){
                           for (int r = 0; r < N; r++){
                          
                                  C[r][c] += (A[r][k] * B[k][c]);
    }}}}
  }
